api/include/opentelemetry/nostd/list.h

```diff
@@ -0,0 +1,254 @@
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cstring>
+#include <iostream>
+#include <iterator>
+#include <memory>
+
+#include "opentelemetry/nostd/utility.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace nostd
+{
+// Non-standard implementation of the list data structure.
+template <class T>
+class list
+{
+public:
+  typedef T value_type;
+
+  // Custom iterator class
+  class iterator
+  {
+  public:
+    iterator operator++()
+    {
+      ptr_++;
+      return ptr_;
+    }
+
+    T &operator*() { return *ptr_; }
+
+    T *operator->() { return ptr_; }
+
+    bool operator==(const iterator &other) { return ptr_ == other.ptr_; }
+
+    bool operator!=(const iterator &other) { return ptr_ != other.ptr_; }
+
+  private:
+    friend class list;
+
+    iterator(T *ptr) : ptr_(ptr) {}
+
+    T *ptr_;
+  };
+
+  list()
+  {
+    size_ = 0;
+    data_ = nullptr;
+  };
+
+  // Constructs class with passed in element.
+  list(T element)
+  {
+    size_ = 1;
+
+    data_ = new T[size_];
+
+    std::memcpy(data_, &element, sizeof(T));
+  };
+
+  // Constructs class with the passed in initializer_list.
+  list(std::initializer_list<T> init_list)
+  {
+    size_ = init_list.size();
+
+    data_ = new T[size_];
+
+    std::memcpy(data_, std::begin(init_list), init_list.size() * sizeof(T));
+  }
+
+  // Returns the element at the front of the list.
+  T &front() { return data_[0]; }
+
+  // Returns the element at the back of the list.
+  T &back()
+  {
+    if (size_ == 0)
+    {
+      return data_[0];
+    }
+
+    return data_[size_ - 1];
+  }
+
+  // Returns an iterator pointing at the front of the list.
+  iterator begin() { return iterator(data_); }
+
+  // Returns an iterator pointing at the back of the list.
+  iterator end() { return iterator(data_ + size_); }
+
+  // Inserts the passed in element at the passed in
+  // iterator position.
+  iterator insert(iterator position, const T &element)
+  {
+    int previous_size = size_;
+    size_++;
+
+    T *temp = new T[size_];
```

Comment:
Can we improve on this? re-allocating for every insert won't be efficient solution.

Reply:
Added a resize function that will reduce the amount of allocations.